File: Source/DebugInfo.cpp
// This is an open source non-commercial project. Dear PVS-Studio, please check it.
// PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
/*
==============================================================================

DebugInfo.cpp

This file is part of MIDI2LR. Copyright 2015 by Rory Jaffe.

MIDI2LR is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

MIDI2LR is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
MIDI2LR.  If not, see <http://www.gnu.org/licenses/>.
==============================================================================
*/
#include "DebugInfo.h"
#include "../JuceLibraryCode/JuceHeader.h"

#ifdef _WIN32
#include <array>
#include "Windows.h"
#include <unicode/unistr.h>

std::string rsj::GetKeyboardLayout()
{
    static_assert(sizeof(WCHAR) == sizeof(UChar),
        "For Unicode handling, assuming windows wide char is same as ICU 16-bit Unicode char.");
    std::array<WCHAR, KL_NAMELENGTH> keyboard_name;
    if (GetKeyboardLayoutNameW(keyboard_name.data())) {
        std::string utf8_rep;
        const icu::UnicodeString utf16_rep{keyboard_name.data()};
        utf16_rep.toUTF8String(utf8_rep);
        return utf8_rep;
    }
    return "unable to get keyboard layout name";
}
#endif


DebugInfo::DebugInfo()
{
    using namespace std::string_literals;
    info.emplace_back("System language "s + juce::SystemStats::getDisplayLanguage().toStdString());
    info.emplace_back("App version "s + ProjectInfo::versionString);
    info.emplace_back("MIDI2LR app path "s +
        juce::File::getSpecialLocation(juce::File::currentApplicationFile).getFullPathName().toStdString());
    info.emplace_back("Keyboard type "s + rsj::GetKeyboardLayout());
}

std::string const * const DebugInfo::GetInfo()
{
    if (iterate >= info.size())
        return nullptr;
    return &info[iterate++];
}

File: src/plugin/DebugInfo.lua
--[[ 

DebugInfo.lua
Saves debug info to a file
 
This file is part of MIDI2LR. Copyright 2015 by Rory Jaffe.

MIDI2LR is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later version.

MIDI2LR is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
MIDI2LR.  If not, see <http://www.gnu.org/licenses/>. 
------------------------------------------------------------------------------]]

local Info           = require 'Info'
local Ut             = require 'Utilities'
local LrApplication  = import 'LrApplication'
local LrFileUtils    = import 'LrFileUtils'
local LrLocalization = import 'LrLocalization'
local LrPathUtils    = import 'LrPathUtils'
local LrSystemInfo   = import 'LrSystemInfo'

local testfile = LrPathUtils.child(_PLUGIN.path,'Client.lua')
local writeable = ''
if not LrFileUtils.isWritable(testfile) then
  writeable = '\nPlugin directory is not writeable'
end

local function gatherInformation()
  local longest = 40
  local mess = '----------- LIGHTROOM -----------'
  ..'\nOperating system '..LrSystemInfo.summaryString()
  ..'\nVersion '..LrApplication.versionString()
  ..'\nLanguage '..LrLocalization.currentLanguage() 
  ..'\nPreferences path '..LrPathUtils.getStandardFilePath ('appPrefs') 
  ..'\nApplication data path '..LrPathUtils.getStandardFilePath ('appData')
  ..'\n----------- PLUGIN -----------' 
  ..'\nVersion '..Info.VERSION.major..'.'..Info.VERSION.minor..'.'..Info.VERSION.revision..'.'..Info.VERSION.build 
  ..'\nPath '.._PLUGIN.path..writeable
  ..'\nMIDI2LR log path '..Ut.applogpath()
  ..'\nMIDI2LR data path '..Ut.appdatapath()
  local lines = 11 -- update if change above message
  if type(Info.AppInfo) == 'table' then
    mess = mess..'\n----------- APP -----------'
    lines = lines + 1
    for _,v in ipairs(Info.AppInfo) do
      longest = math.max(longest, v:len())
      lines = lines + 1
      mess = mess..'\n'..v
    end
  end
  return mess, lines, longest
end

local function writeDebug()
  local datafile = LrPathUtils.child(Ut.applogpath(), 'MIDI2LRinfo.txt')
  local file = assert(io.open(datafile,'w'),LOC("$$$/AgImageIO/Errors/WriteFile=The file could not be written.")..' '..datafile)
  file:write(gatherInformation())
  file:close()
end

local function sendLog()
  MIDI2LR.SERVER:send('Log Operating system '..LrSystemInfo.summaryString()..'\n')
  MIDI2LR.SERVER:send('Log Lightroom Version '..LrApplication.versionString()..'\n')
  MIDI2LR.SERVER:send('Log Lightroom Language '..LrLocalization.currentLanguage()..'\n')
  MIDI2LR.SERVER:send('Log Lightroom Preferences path '..LrPathUtils.getStandardFilePath ('appPrefs')..'\n')
  MIDI2LR.SERVER:send('Log Lightroom Application data path '..LrPathUtils.getStandardFilePath ('appData')..'\n')
  MIDI2LR.SERVER:send('Log Plugin version '..Info.VERSION.major..'.'..Info.VERSION.minor..'.'..Info.VERSION.revision..'.'..Info.VERSION.build..'\n')
  MIDI2LR.SERVER:send('Log Plugin path '.._PLUGIN.path..writeable..'\n')
  MIDI2LR.SERVER:send('Log Plugin log path '..Ut.applogpath()..'\n')
  MIDI2LR.SERVER:send('Log Plugin data path '..Ut.appdatapath()..'\n')
end

return {
  gatherInformation = gatherInformation,
  sendLog = sendLog,
  write = writeDebug,
}